File: 第一阶段/1.3-基本数据类型/BasicTypes.cs
// C# 基本数据类型示例
// 对应文章：1.3 基本数据类型

using System;
using System.Numerics;

// ==================== 整数类型（多种精度！） ====================
Console.WriteLine("=== 整数类型 ===");
byte a = 255;                    // 8位，0-255
short b = 32767;                 // 16位
int c = 2147483647;              // 32位，最常用
long d = 9223372036854775807;    // 64位

Console.WriteLine($"byte:   {a} (8位)");
Console.WriteLine($"short:  {b} (16位)");
Console.WriteLine($"int:    {c} (32位，最常用)");
Console.WriteLine($"long:   {d} (64位)");

// C# 没有无限精度整数，需要 BigInteger
BigInteger big = BigInteger.Pow(2, 1000);
Console.WriteLine($"2^1000 的位数: {big.ToString().Length}");  // 302 位！

// 不同进制表示
Console.WriteLine($"二进制:  {Convert.ToInt32("1010", 2)}");   // 10
Console.WriteLine($"十六进制: 0x{255:X}");                      // 0xFF

// ==================== 浮点数 ====================
Console.WriteLine("\n=== 浮点数 ===");
float e = 3.14f;       // 32位单精度（注意 f 后缀！）
double f = 3.14;       // 64位双精度，最常用
decimal g = 3.14m;     // 128位，金融计算（注意 m 后缀！）

Console.WriteLine($"float:   {e} (32位)");
Console.WriteLine($"double:  {f} (64位，最常用)");
Console.WriteLine($"decimal: {g} (128位，金融计算)");

// 浮点数精度问题
Console.WriteLine($"0.1 + 0.2 = {0.1 + 0.2}");          // 0.30000000000000004
Console.WriteLine($"0.1 + 0.2 == 0.3? {0.1 + 0.2 == 0.3}");  // False!
// C# 的 decimal 不会有这个问题
Console.WriteLine($"decimal: {0.1m + 0.2m == 0.3m}");    // True

// ==================== 布尔类型 ====================
Console.WriteLine("\n=== 布尔类型 ===");
bool a1 = true;
bool b1 = false;
Console.WriteLine($"true && false = {true && false}");  // false
Console.WriteLine($"true || false = {true || false}");  // true
Console.WriteLine($"!true         = {!true}");          // false

// C# 的 bool 不能当数字用！
// int x = (int)true;  // 编译错误！需要显式转换
Console.WriteLine($"(int)true = {(int)true}");           // 1（需要显式转换）
Console.WriteLine($"(int)false = {(int)false}");         // 0

// ==================== 字符串类型 ====================
Console.WriteLine("\n=== 字符串类型 ===");
string s1 = "Hello";
string s2 = "World";
string s3 = s1 + " " + s2;     // 拼接
string s4 = $"{s1} {s2}";      // 插值（推荐）
Console.WriteLine($"拼接: {s3}");
Console.WriteLine($"插值: {s4}");

// 字符串是不可变的
int len = s1.Length;            // 5（属性，不是函数！Python 用 len()）
string upper = s1.ToUpper();   // "HELLO"
string sub = s1.Substring(1, 3);  // "ell"
Console.WriteLine($"长度: {len}, 大写: {upper}, 子串: {sub}");

// ==================== 空值 ====================
Console.WriteLine("\n=== 空值 ===");
string s = null;
int? x = null;            // 可空类型（C# 独有！Python 没有）
double? y = null;
Console.WriteLine($"s == null: {s == null}");           // true
Console.WriteLine($"x has value: {x.HasValue}");       // false
Console.WriteLine($"x ?? default: {x ?? 0}");          // 0（空合并运算符）

File: 第一阶段/1.3-基本数据类型/basic_types.py
# Python 基本数据类型示例
# 对应文章：1.3 基本数据类型

# ==================== 整数类型（统一的 int，无上限！） ====================
print("=== 整数类型 ===")
a = 255
b = 32767
c = 2147483647
d = 9223372036854775807
# Python 的 int 没有上限，随便算！
e = 2 ** 1000
print(f"2^1000 的位数: {len(str(e))}")  # 302 位！C# 需要 BigInteger

# 不同进制表示
print(f"二进制:  {0b1010}")      # 10
print(f"八进制:  {0o17}")        # 15
print(f"十六进制: {0xFF}")       # 255

# ==================== 浮点数 ====================
print("\n=== 浮点数 ===")
f1 = 3.14       # 64位双精度（Python 没有 float32/decimal 区分）
f2 = 3.14e10    # 科学计数法 = 3.14 × 10^10
f3 = 1.5e-4     # 0.00015
print(f"3.14e10 = {f2}")

# 浮点数精度问题（所有语言都有！）
print(f"0.1 + 0.2 = {0.1 + 0.2}")         # 0.30000000000000004
print(f"0.1 + 0.2 == 0.3? {0.1 + 0.2 == 0.3}")  # False！
# 精确计算用 decimal 模块
from decimal import Decimal
print(f"Decimal: {Decimal('0.1') + Decimal('0.2')}")  # 0.3

# ==================== 复数（C# 没有的！） ====================
print("\n=== 复数 ===")
h = 3 + 4j     # Python 原生支持复数
print(f"复数: {h}")
print(f"实部: {h.real}")    # 3.0
print(f"虚部: {h.imag}")    # 4.0
print(f"模: {abs(h)}")      # 5.0（勾股定理）

# ==================== 布尔类型（就是整数！） ====================
print("\n=== 布尔类型 ===")
a1 = True
b1 = False
print(f"True and False = {True and False}")  # False
print(f"True or False  = {True or False}")   # True
print(f"not True       = {not True}")        # False

# 布尔就是整数！（C# 做不到！）
print(f"\nTrue + True     = {True + True}")     # 2
print(f"False * 10      = {False * 10}")       # 0
print(f"True * 5 + False = {True * 5 + False}") # 5
print(f"sum([True, True, False, True]) = {sum([True, True, False, True])}")  # 3

# ==================== 字符串类型 ====================
print("\n=== 字符串类型 ===")
s1 = "Hello"
s2 = 'World'         # 单引号和双引号一样
s3 = s1 + " " + s2   # 拼接
s4 = f"{s1} {s2}"    # f-string（推荐，Python 3.6+）
print(f"拼接: {s3}")
print(f"f-string: {s4}")

# 字符串是不可变的序列
length = len(s1)           # 5（函数，不是属性！C# 用 .Length）
upper = s1.upper()         # "HELLO"
sub = s1[1:4]              # "ell"（切片语法）
print(f"长度: {len('Hello')}, 大写: {'hello'.upper()}, 切片: {'Hello'[1:4]}")

# ==================== 空值 None ====================
print("\n=== 空值 None ===")
s = None
x = None
print(f"s is None: {s is None}")   # True（推荐用 is 判断 None）
print(f"s is not None: {s is not None}")  # False
# C# 对比：None ≈ null，但 None 是单例对象
# 判断用 is 不用 ==，因为 == 可能被 __eq__ 重载
